Fix bug where pattern-matching treats a type parameter as if it is a reference (which it might not be)

gafter opened the pull request:
Fixes #35584

@agocke @cston @jcouv for review.  This is a serious bad-code-gen regression and should be fixed in 16.1.  I'm not certain that I am targeting the correct branch.

**Customer and scenario info**

**Who is impacted by this bug?**
Customers who use pattern-matching in C# 7 in VS2019.

**What is the customer scenario and impact of the bug?**
The scenario is a pattern-matching operation where the input is a type parameter and it is being matched into a variable of type `object` or some reference base type of the type parameter.  The compiler produces unverifiable code, omitting a necessary `box` instruction on the input.  This causes a runtime crash when the type parameter is instantiated with a value type.

**What is the workaround?**
None

**How was the bug found?**
Customer-reported

**If this fix is for a regression - what had regressed, when was the regression introduced, and why was the regression originally missed?**

This is a regression in VS2019 versus previous VS2017.  We rewrote the code generation for pattern-matching from scratch in VS2019 in order to add other (experimental for C# 8) pattern-matching features, and this particular interaction was missed in testing.  The test gap is fixed here.

agocke commented on src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs:
```diff
@@ -2385,5 +2389,223 @@ .locals init (bool V_0)
+    public static void Main()
+    {
+        var nonGenericfoo = new NonGenericFoo(0);  // no problem
+        var genericFoo = new GenericFoo<int>(0); // Kaboom exception thrown
```
Why would this throw? I also don't see it throwing in the output

gafter replied:
It doesn't throw because the bug is fixed.  Should I revise the comments?

thasc replied:
Makes sense - seems like comment rot otherwise. They were only added to call out where the bug occurred, so now that it no longer occurs, they're outdated, and their presence in the test code suggests that an exception is actually expected. I'd suggest either removing them, or adjusting them to indicate that a bug *used* to occur there.

agocke replied:
Yup, this reads very strangely right now

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs:
```diff
@@ -335,20 +337,38 @@ private BoundExpression MakeEqual(BoundExpression loweredLiteral, BoundExpressio
+                    BoundExpression output = _tempAllocator.GetTemp(new BoundDagTemp(evaluation.Syntax, baseType, evaluation));
+                    sideEffect = _factory.AssignmentExpression(output, _factory.Convert(baseType, input));
+                    testExpression = _factory.ObjectNotEqual(output, _factory.Null(baseType));
+                    _localRewriter._diagnostics.Add(test.Syntax, useSiteDiagnostics);
```
>_diagnostics.Add(test.Syntax, useSiteDiagnostics) [](start = 35, length = 49)

Are we testing use-site diagnostics?

gafter replied:
It is not possible to get here without the use site diagnostics being reported in binding (we are checking for a conversion between the input type and the type being matched), which means we probably cannot get here (in lowering) with a use site diagnostic.  But I'm reporting them in case there is something I didn't think of.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs:
```diff
@@ -2385,5 +2389,223 @@ .locals init (bool V_0)
+        var genericFoo = new GenericFoo<int>(0); // Kaboom exception thrown
+    }
+}
+public class NonGenericFoo
```
>Foo [](start = 23, length = 3)

Please remove.

gafter commented:
I think I've addressed all comments.  @agocke @cston do you have any further comments?  Please let me know if you'd like an explanation of the bug fix.

gafter commented:
@jcouv Do you know if this is the correct branch?  What is my next step?

jcouv commented:
Approved in email with Art and Jinu. I'll go ahead and merge.
